include/boost/gil/image_processing/morphology.hpp

```diff
@@ -0,0 +1,193 @@
+            {
+                dilate,
+                erode,
+                opening,
+                closing,
+                morphological_gradient,
+                top_hat,
+                black_hat
+            };
+namespace detail{
+template <typename SrcView, typename DstView, typename Kernel>
+void morph_impl(SrcView const& src_view, DstView const& dst_view, Kernel const& kernel,const int identifier)
+{
+    int flip_ker_row, flip_ker_col, row_boundary, col_boundary;
+    typename channel_type<typename SrcView::value_type>::type max_overlapped_element,min_overlapped_element;
+    for (std::ptrdiff_t view_row = 0; view_row < src_view.height(); ++view_row)
+    {
+        for (std::ptrdiff_t view_col = 0; view_col < src_view.width(); ++view_col)
+        {
+            max_overlapped_element = (std::numeric_limits<typename channel_type<SrcView>::type>::min)();
+            min_overlapped_element = (std::numeric_limits<typename channel_type<SrcView>::type>::max)();
+            for (std::size_t kernel_row = 0; kernel_row < kernel.size(); ++kernel_row)
+            {
+                flip_ker_row = kernel.size() - 1 - kernel_row;      // row index of flipped kernel
+
+                for (std::size_t kernel_col = 0; kernel_col < kernel.size(); ++kernel_col)
+                {
+                    flip_ker_col = kernel.size() - 1 - kernel_col; // column index of flipped kernel
+
+                    // index of input signal, used for checking boundary
+                    row_boundary = view_row + (kernel.center_y() - flip_ker_row);
+                    col_boundary = view_col + (kernel.center_x() - flip_ker_col);
+
+                    // ignore input samples which are out of bound
+                    if (row_boundary >= 0 && row_boundary < src_view.height() &&
+                        col_boundary >= 0 && col_boundary < src_view.width())
+                    {
+                        if(src_view(col_boundary, row_boundary) > max_overlapped_element)
+                            max_overlapped_element = src_view(col_boundary, row_boundary);
+                        if(src_view(col_boundary, row_boundary) < min_overlapped_element)
+                            min_overlapped_element = src_view(col_boundary, row_boundary);
+                    }
+                }
+            }
+            if(identifier)//identifier = 1 for dilation
+                dst_view(view_col, view_row) = max_overlapped_element;
+            else //identifier = 0 for erosion
+                dst_view(view_col, view_row) = min_overlapped_element; 
+        }
+        }
+    }
+
+template <typename SrcView, typename DstView,typename Kernel>
+void morph(SrcView const& src_view, DstView const& dst_view,Kernel const& ker_mat,const int identifier)
+{
+    using namespace boost::gil;
+    BOOST_ASSERT(ker_mat.size() != 0 && src_view.dimensions() == dst_view.dimensions());
+    gil::gray8_image_t intermediate_img(src_view.dimensions());
```

Comment:
Do we need this?

Reply:
Yes I think we need this,dilate function has
```
for(int i=0;i<iterations;++i)
        morph(dst_view,dst_view,ker_mat,1);
```
so I am passing dst_view as input view as well as output view in morph ,if I do not create a dummy/intermediate image to pass in morph_impl,modified values of dst_view will be used to modify new values in dst_view,this will not output the result of typical dilation.Moreover,if i choose to pass a separate image view than dst_view in dilate,then i would have to create new image views for each and every iteration which seems like a bad idea from the point of view of space complexity. I understand that all images will not be of the type gray8_image_t and I am thinking of a way to solve it.Feel free to share what you think of it

Reply:
@meshtag
> I am passing dst_view as input view as well as output view in morph

I read this and it immediately smells to me, sorry.
First, we use `dst_` as convention for output-only, a result destination.
Second, how `T const& v` can be an output?

Reply:
@mloskot 
1.Regarding the first point,okay I will be more careful about the terminologies/conventions used from now on.Thanks for pointing that out.
2.Regarding the second point,documentation states that gil views are similar to standard iterators(pointers),so when I am writing ```T const& v ``` where v is a gil view,it should mean that I am using a constant pointer and not a pointer with constant value,this allows me to fill dst_view . Please correct me if I am wrong.

Reply:
The 1. is simply about striving for naming things as clearly as possible.

Yes, you are correct about the 2. and the [docs states it quite clearly indeed](https://www.boost.org/doc/libs/1_75_0/libs/gil/doc/html/tutorial/gradient.html): 

> ```
> void x_gradient(gray8c_view_t const& src, gray8s_view_t const& dst)
> ```
> ...
>  Whether a view is mutable or read-only (immutable) is a property of the view type.

I'm sorry, I must be thinking of something else last night.

Well, I can also see this may be confusing to those new to GIL and those used to understanding `T const&` as read-only stuff. Since,

> A GIL image view, is a shallow, lightweight view of a rectangular grid of pixels.

we could do just:

```cpp
void x_gradient(gray8c_view_t input, gray8s_view_t output)
```

Something to think about for future readability and code quality refactoring in GIL :-)

Reply:
@mloskot 
```gray8c_view_t``` and ```gray8s_view_t``` specify what we can do to the image pixels,but I am not sure if they specify whether the entity to which the view(pointer/iterator) is pointing will remain constant or not?,what if after passing those views,someone tries to modify them (possibly by operator overloading) like this ```dst_view += 1```(the way we do it with a normal pointer) inside the function,will our compiler shout at them to stop it because modifying to what the view is pointing inside a function would certainly create problems, I would look into the matter more closely once I finish working on this pull request.The above problem is certainly solved if we pass the view in the current manner ```DstView const& dst_view``` since it would create something like a constant pointer,I need to do a lot of testing before I arrive on a conclusion,as I said I will invest time on it once I finish my work on this pr.

Reply:
> the view(pointer/iterator like entity)

Initially, it is not a major error to think of view as pointer or iterator, but at this stage I'd suggest to make the clear distinction.
There used to be a great article by Maciej Sobczak archived online "STL Sequences & the View Concept", C/C++ Users Journal, April 2004 Vol. 22, No. 4 but it seems no longer online. It explained the concept of a view very well. You may look at the docs of https://www.zib.de/weiser/vtl/ to grasp the idea.

Essentially, a view is a container adaptor that allows us to 'see' a container in certain way - think of an eyewear.
So, it is not really equivalent of a pointer or iterator, but view interacts with pointers or iterators.
In GIL, view provides access to iterators and locators (another GIL concept) over the underlying sequence of pixels.

> what if after passing the view,someone tries to modify it 

Let's try https://godbolt.org/z/vbdne5

```cpp
#include <boost/gil.hpp>
namespace gil = boost::gil;

int main()
{
    gil::gray8_view_t rw;
    auto rw_pixel = rw(1, 1);
    rw(1, 1) = 128;

    gil::gray8c_view_t ro;
    auto ro_pixel = ro(1, 1);
    ro(1, 1) = 128; // and?
}
```

By the way, from `04:00` minute the [GIL Video Lecture](https://www.youtube.com/watch?v=sR8Wjg0pceE&t=241) explains what is the `c` in `gray8c_view_t`, what is `s` in `gray8s_view_t`, etc.

> what if after passing the view,someone tries to modify it (possibly by operator overloading) like this `dst_view += 1` (the way we do it with a normal pointer) inside the function

Because the `image_view` is **not** an iterator or pointer, there is nothing in it to modify like that, and there is no such `operator+=`. That is what makes the view lightweight too. The `image_view` class has only two members `dimensions` and `pixels` and there is no way you can modify it apart from assigning one view to another what will completely reset the two resulting in a completely new view.

Reply:
Okay I think I get the difference between iterators and views now.Thanks a lot @mloskot  for helping me improve my understanding of this concept .